mlir/lib/Catalyst/Driver/CompilerDriver.cpp

```diff
@@ -0,0 +1,443 @@
+                                   Type assumedElementType,
+                                   SmallVectorImpl<RankedTensorType> &inferredTypes)
+{
+    auto inferSingleMemRef = [&](llvm::StructType *descriptorType) {
+        SmallVector<int64_t> resultShape;
+        assert(descriptorType->getNumElements() >= 3 &&
+               "Expected MemRef descriptor struct to have at least 3 entries");
+        if (descriptorType->getNumElements() == 3) {
+            // resultShape is empty
+        }
+        else {
+            auto *arrayType = cast<llvm::ArrayType>(descriptorType->getTypeAtIndex(3));
+            size_t rank = arrayType->getNumElements();
+            for (size_t i = 0; i < rank; i++) {
+                resultShape.push_back(ShapedType::kDynamic);
+            }
+        };
+        return RankedTensorType::get(resultShape, assumedElementType);
+    };
+    if (returnType->isVoidTy()) {
+        return failure();
+    }
+    if (auto *structType = dyn_cast<llvm::StructType>(returnType)) {
+        // The return type could be a single memref descriptor or a struct of multiple memref
+        // descriptors.
+        if (isa<llvm::StructType>(structType->getElementType(0))) {
+            for (size_t i = 0; i < structType->getNumElements(); i++) {
+                inferredTypes.push_back(
+                    inferSingleMemRef(cast<llvm::StructType>(structType->getTypeAtIndex(i))));
+            }
+        }
+        else {
+            // Assume the function returns a single memref
+            inferredTypes.push_back(inferSingleMemRef(structType));
+        }
+        return success();
+    }
+    return failure();
+}
+
+LogicalResult runLLVMPasses(const CompilerOptions &options,
+                            std::shared_ptr<llvm::Module> llvmModule,
+                            CompilerOutput::PipelineOutputs &outputs)
+{
+    // opt -O2
+    // As seen here:
+    // https://llvm.org/docs/NewPassManager.html#just-tell-me-how-to-run-the-default-optimization-pipeline-with-the-new-pass-manager
+
+    // Create the analysis managers.
+    llvm::LoopAnalysisManager LAM;
+    llvm::FunctionAnalysisManager FAM;
+    llvm::CGSCCAnalysisManager CGAM;
+    llvm::ModuleAnalysisManager MAM;
+    // Create the new pass manager builder.
+    // Take a look at the PassBuilder constructor parameters for more
+    // customization, e.g. specifying a TargetMachine or various debugging
+    // options.
+    llvm::PassBuilder PB;
+    // Register all the basic analyses with the managers.
+    PB.registerModuleAnalyses(MAM);
+    PB.registerCGSCCAnalyses(CGAM);
+    PB.registerFunctionAnalyses(FAM);
+    PB.registerLoopAnalyses(LAM);
+    PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
+    // Create the pass manager.
+    // This one corresponds to a typical -O2 optimization pipeline.
+    llvm::ModulePassManager MPM = PB.buildPerModuleDefaultPipeline(llvm::OptimizationLevel::O2);
+
+    // Optimize the IR!
+    MPM.run(*llvmModule.get(), MAM);
```

Comment:
Awesome!

Reply:
For my own curiosity, are these passes are standard O2 optimization analysis? Is there any missing passes we need to consider registering in the future?

Reply:
```c++
    // opt -O2
    // As seen here:
    // https://llvm.org/docs/NewPassManager.html#just-tell-me-how-to-run-the-default-optimization-pipeline-with-the-new-pass-manager
```

The comment above says that this is the way to run the default optimization pipeline. Asking if these are the standard O2 optimization passes is perhaps a bit of a bad question, because it may be possible for different compiler drivers to have different O2 optimizations. E.g., `opt -O2` may be different from `clang -O2` or `clang++ -O2` and of course different from `gcc -O2`. So, are these the "standard O2 optimization analysis"? Yes, if you define the standard as those applied by `opt`. I think most should be the same between `clang -O2` and `opt -O2` but I am not 100% sure how to query for which optimizations are applied at different optimization levels for clang. In gcc is simple: `gcc -Q --help=optimizers -O${N}`.

> Is there any missing passes we need to consider registering in the future?

We are only running these because Enzyme thesis is that it is best to optimize, apply auto differentiation, and optimize again. At the moment I do not think we are considering adding more passes in the future, but we can certainly add more if we have the need. :)